### PR TITLE
Support multipart range requests

### PR DESCRIPTION
**Pull request by fatkodima**
Closes #559

### CODE REVIEW

File: lib/rack/files.rb [hunk 1]
```diff
@@ -70,75 +71,111 @@ def serving(request, path)
...
+        status = 206
         range = ranges[0]
-        response[0] = 206
-        response[1]["Content-Range"] = "bytes #{range.begin}-#{range.end}/#{size}"
+        headers["Content-Range"] = "bytes #{range.begin}-#{range.end}/#{size}"
         size = range.end - range.begin + 1
+      else
+        # Partial content, multiple ranges:
+        status = 206
+        headers[CONTENT_TYPE] = "multipart/byteranges; boundary=#{MULTIPART_BOUNDARY}"
```
Comment (ioquatix) [inline lib/rack/files.rb]: Sorry, I only have a short time to review this, but I'm excited to see this PR.

Is `AaB03x` sufficient as a boundary?

File: lib/rack/files.rb [hunk 2]
```diff
@@ -70,75 +71,111 @@ def serving(request, path)
...
-      if request.head?
-        []
-      else
-        Iterator.new path, range
+      def multipart?
+        ranges.size > 1
+      end
+
+      def multipart_heading(range)
+<<-EOF
```
Comment (ioquatix) [inline lib/rack/files.rb]: Can we encode this as a string rather than a heredoc.

Comment (fatkodima) [reply lib/rack/files.rb]: I tried to stick here with existing code styles, like
https://github.com/rack/rack/blob/93dfcdf46084760079514c858ae7391e90a7821e/lib/rack/multipart/generator.rb#L71-L89

and like examples in many test cases. 
So should I change this to string?

File: lib/rack/files.rb (hunk 1 above)
Comment (fatkodima) [reply lib/rack/files.rb]: Yes, why not? The same boundary is used for multipart forms https://github.com/rack/rack/blob/93dfcdf46084760079514c858ae7391e90a7821e/lib/rack/multipart.rb#L14

**Comment by jeremyevans** [comment]
@fatkodima I think the implementation looks pretty good.  I think you can keep using the existing static boundary. I also think you can keep using the heredoc, you don't need to change it to a string.  Can you please rebase this against master?  Please make sure that it works for multiple ranges the same as for single ranges (which now use `BaseIterator`, see 3936b90697fcaf02751e596bd9dfd0e24add9b29).

**Comment by fatkodima** [comment]
@jeremyevans Updated.

**Comment by jeremyevans** [comment]
Looks good.  I tested locally with `curl -r` and got the expected results.  I plan on merging in a few days unless other committer objects.

**Pull request merged**